### PR TITLE
Clarify the default `cargo run` error message on a non-zero exit code

### LINKED ISSUE #1516: `cargo run` error output is vague and more verbose than with --verbose
There is an issue with cargo

To learn more, please ask me again with `--verbose`.

### PR DESCRIPTION
**Pull request by gkoz**
Closes #1516

### CODE REVIEW

**Comment by gkoz** [comment]
[This spurious error](https://travis-ci.org/rust-lang/cargo/jobs/85029566#L806) might be worth looking into.

**Comment by alexcrichton** [comment]
It does indeed [look spurious](https://github.com/rust-lang/cargo/pull/2055#issuecomment-147786516)!

File: src/bin/run.rs [hunk 1]
```diff
@@ -92,7 +92,10 @@ pub fn execute(options: Options, config: &Config) -> CliResult<Option<()>> {
         None => Ok(None),
         Some(err) => {
             Err(match err.exit.as_ref().and_then(|e| e.code()) {
-                Some(i) => CliError::from_error(err, i),
+                Some(code) => {
+                    let desc = format!("Process finished with exit status {}", code);
+                    CliError::from_boxed(human(desc), code)
```
Comment (alexcrichton) [inline src/bin/run.rs]: Hm unfortunately this is throwing away `err` which has lots of other contextual information which shouldn't be lost. I'm also a little concerned about displaying this sort of information by default. I see `cargo run` as a "build stuff and then exec" in the sense that the process basically "becomes" the child.

**Comment by gkoz** [comment]
> Hm unfortunately this is throwing away err which has lots of other contextual information which shouldn't be lost.

I could change this to chain the error. I didn't go with chaining because I'm not sure what those lots of information are and the outcome would be "the process returned 2, pass `-v` for more details" and `-v` would tell you "the process returned 2, caused by: the process (path) returned 2". This seems silly.

> I'm also a little concerned about displaying this sort of information by default. I see cargo run as a "build stuff and then exec" in the sense that the process basically "becomes" the child.

I don't mind cargo "becoming" the child but in that case it's not cargo's business to proclaim a non-zero exit code to be an error, more so an "unknown error". This makes `cargo run`ning console apps a bad experience (is `test 1 = 2` returning `1` an error?). A decision is long overdue: either give a useful error message or remove it altogether (not sure how this would work).

**Comment by alexcrichton** [comment]
Oh right we print things out for errors no matter what, in that case this is basically just the same as the contained error so I think it'd be best to just convert the error into a human one where it's actually generated and then that'll propagate upwards.

**Comment by gkoz** [comment]
If we return `Err` from where the error is generated, the exit status will get buried in the `Box`  requiring downcasts to get it out. Currently this is worked around by [returning `Ok` with an error inside](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_run.rs#L57) and I believe I'm doing the best I can with that here.

**Comment by gkoz** [comment]
Or do you mean changinh `Option<ProcessError>` to `Option<Human<ProcessError>>` [here](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_run.rs#L9)?

**Comment by gkoz** [comment]
Anyway [the function I'm changing](https://github.com/gkoz/cargo/blob/run_human_error/src/bin/run.rs#L51) seems required to return `CliResult` because of macros that implement the subcommands.

**Comment by alexcrichton** [comment]
The API function probably doesn't want to be changed as it's returning specifically a `ProcessError`, but instead of using the `human` function you could create a `Human` error, box that up into a `CargoError` and then pass that to `CliError`

**Comment by gkoz** [comment]
I've wrapped the `ProcessError` in `Human` but now the default mode prints too much information:
```text
Process didn't exit successfully: `target/debug/foo` (exit code: 2)
```

**Comment by gkoz** [comment]
If I were to use `caused_human` instead the output would be
```text
Process finished with exit status 2

To learn more, run the command again with --verbose.
```
Verbose:
```text
Process finished with exit status 2

Caused by:
  Process didn't exit successfully: `target/debug/foo` (exit code: 2)
```

This is the silliness I referred to earlier. More verbosity and confusion.

What kind of information exactly was I throwing away originally?

**Comment by alexcrichton** [comment]
Ah yeah this is what I was thinking, but you feel this is too much information? Is that because the path of the executable is printed out?

**Comment by gkoz** [comment]
I believe printing the path does not fall in the "useful to see by default" category but I don't have strong feeling about this. Also the statement that the process didn't exit successfully, which is not universally true, remains.

In the original proposal the verbose message would be missing the path (it would be apparent from the `Running ...` message though) and the default message would avoid it nicely.

Generally, I'm fine with both.

**Comment by alexcrichton** [comment]
@bors: r+ 52553c981e21bfe1724b20718a69517ab3dec565

OK, I think that this is better than what we have right now and it's still not inundating you with too much information, so I'm fine with this. We can perhaps continue to refine it down the road! For now I've also edited the PR description to close the associated issue.

**Comment by gkoz** [comment]
Thanks!

**Pull request merged**